model/policy.go

```diff
@@ -123,34 +124,37 @@ func (model Model) HasPolicies(sec string, ptype string, rules [][]string) bool
 // AddPolicy adds a policy rule to the model.
 func (model Model) AddPolicy(sec string, ptype string, rule []string) {
 	model[sec][ptype].Policy = append(model[sec][ptype].Policy, rule)
+	model[sec][ptype].PolicyMap[strings.Join(rule, DefaultSep)] = len(model[sec][ptype].Policy) - 1
 }
 
 // AddPolicies adds policy rules to the model.
 func (model Model) AddPolicies(sec string, ptype string, rules [][]string) {
-	for i := 0; i < len(rules); i++ {
-		model[sec][ptype].Policy = append(model[sec][ptype].Policy, rules[i])
+	for _, rule := range rules {
+		model.AddPolicy(sec, ptype, rule)
 	}
 }
 
 // RemovePolicy removes a policy rule from the model.
 func (model Model) RemovePolicy(sec string, ptype string, rule []string) bool {
-	for i, r := range model[sec][ptype].Policy {
-		if util.ArrayEquals(rule, r) {
-			model[sec][ptype].Policy = append(model[sec][ptype].Policy[:i], model[sec][ptype].Policy[i+1:]...)
-			return true
-		}
+	index, ok := model[sec][ptype].PolicyMap[strings.Join(rule, DefaultSep)]
+	if !ok {
+		return false
 	}
 
-	return false
+	model[sec][ptype].Policy = append(model[sec][ptype].Policy[:index], model[sec][ptype].Policy[index+1:]...)
+	delete(model[sec][ptype].PolicyMap, strings.Join(rule, DefaultSep))
```

Comment:
concurrent delete map element will panic

Reply:
Caspin should use thread safe instance by default.

Reply:
I think the issue of thread safety should be provided by SyncedEnforcer. Before, the model was not thread safe. Even if we guarantee that map is thread-safe, policy slice still requires addition another lock

Reply:
@dovics you're right.